### DISCUSSION TITLE
Handling Error with appstate

### DISCUSSION THREAD

**Discussion by 0xpr03** [tokio-rs/axum#2272] [Q&A] [answered]
How do you typically handle the following requirements in axum?

- early-return in handlers when no valid login could be found
- use some kind of IntoResponse AppError
- render that error for all handlers with a template system that requires access to the AppState and also possibly more context (for example the Url when redirecting towards a login that should also send you back to your original page after logging in).

It currently looks like I would have to use a `HandleError` for every handler, creating a ton of `.route("foo",HandleError(error_handler,app_logic))`. Something like the generalized HandleErrorLayer (which also has state access) would be nice here - but obviously won't work for my non-service. (is it even a good idea to make every route call a service?)

**Comment by davidpdrsn** [comment, marked as answer]
> early-return in handlers when no valid login could be found

Make your own extractor or middleware.

> use some kind of IntoResponse AppError

IntoResponse intentionally doesn’t access to state so you have to pass that in some other way.

> render that error for all handlers with a template system that requires access to the AppState and also possibly more context 

This also sounds like a middleware to me. Perhaps the handler sets a response extension which is picked up by a middleware and then converted into a response.

**Reply by 0xpr03** [reply]
Abstracting the login-requirements with a middleware probably is a good idea - I just feared creating my own. Using another one for error handling doesn't sound bad, then I could even keep my IntoResponse Error-Enum.

**Reply by davidpdrsn** [reply]
It’s pretty easy to make a middleware with https://docs.rs/axum/0.6.20/axum/middleware/fn.from_fn.html
